crates/dbsp/src/tutorial.rs

```diff
@@ -0,0 +1,744 @@
+//!                 },
+//!             )
+//!         })
+//! ```
+//!
+//! Getting the top three months for each country and printing them is
+//! then as simple as:
+//!
+//! ```ignore
+//!         .topk_desc(3)
+//!         .inspect(|data| {
+//!             data.iter()
+//!                 .for_each(|(c, VaxMonthly { count, year, month }, w)| {
+//!                     println!("{c:16} {year}-{month:02} {count:10}: {w:+}")
+//!                 })
+//!         });
+//! ```
+//!
+//! The complete program is in `tutorial8.rs`.  When we run it, it outputs:
+//!
+//! ```text
+//! England          2021-03   11861175: +1
+//! England          2021-05   13772946: +1
+//! England          2021-12   14801300: +1
+//! Northern Ireland 2021-05     394047: +1
+//! Northern Ireland 2021-04     436870: +1
+//! Northern Ireland 2021-12     489059: +1
+//! Scotland         2021-06    1244155: +1
+//! Scotland         2021-05    1272194: +1
+//! Scotland         2021-12    1388549: +1
+//! Wales            2021-04     707714: +1
+//! Wales            2021-12     822945: +1
+//! Wales            2021-03     836844: +1
+//! ```
+//!
+//! ### Vaccination rates
+//!
+//! Suppose we want to compare the population in each country with the
+//! number of vaccinations given, that is, to calculate a vaccination
+//! rate.  Our input data contains a `total_vaccinations_per_hundred`
+//! column that reports this information, but let's calculate it ourselves
+//! to demonstrate how it might be done in DBSP.
+//!
+//! We need to know the population in each country.  That's a separate
+//! input table.  Obviously, population changes over time, but for
+//! simplicity let's just pick a single figure for each one.
+//!
+//! Up until now, our circuit only had a single input source.  We'll add a
+//! second one for population by country.  Since that's naturally a set of
+//! key-value pairs, we'll make it an indexed Z-set input source instead of a
+//! plain Z-set.  The first step is to make our circuit builder construct the
+//! source and return its handle:
+//!
+//! ```ignore
+//! fn build_circuit(
+//!     circuit: &mut RootCircuit,
+//! ) -> Result<(
+//!     CollectionHandle<Record, isize>,
+//!     CollectionHandle<String, (u64, isize)>,
+//! )> {
+//!     let (vax_stream, vax_handle) = circuit.add_input_zset::<Record, isize>();
+//!     let (pop_stream, pop_handle) = circuit.add_input_indexed_zset::<String, u64, isize>();
+//! ...
+//!     Ok((vax_handle, pop_handle))
+//! }
+//! ```
+//!
+//! The client code needs to receive the additional handle and feed data into
+//! it.  Let's feed in fixed data this time:
+//!
+//! ```ignore
+//!     let (circuit, (mut vax_handle, mut pop_handle)) = RootCircuit::build(build_circuit)?;
+//! ...
+//!     let mut pop_records = vec![
+//!         ("England".into(), (56286961, 1)),
+//!         ("Northern Ireland".into(), (1893667, 1)),
+//!         ("Scotland".into(), (5463300, 1)),
+//!         ("Wales".into(), (3152879, 1)),
```

Comment:
Btw, another way to demonstrate incrementality is to modify some of these records later in the tutorial.